core/variant/reduz_struct_proposal.h

```diff
@@ -0,0 +1,255 @@
+	}
+
+	void Array::set_struct_member_by_offset(uint32_t p_offset,const Variant& p_value) {
+	ERR_FAIL_COND(!_p->struct_layout.is_struct());
+	int32_t offset = p_offset;
+	offset += struct_offset * _p->struct_size;
+	ERR_FAIL_INDEX(offset,_p->array.size());
+	_p->array[offset].write[offset]=p_value;
+	}
+
+
+	// TYPE DESCRIPTIONS in C++
+
+	// Another problem we will face with this approach is that there are many cases where we will want to actually describe the type.
+	// If we had a function that returned a dictionary and now we want to change it to a struct because its easier for the user to use (description in doc, autocomplete in GDScript, etc) we must find a way. As an example for typed arrays we have:
+
+	TypedArray<Type> get_someting() const;
+
+	// And the binder takes care. Ideally we want to be able to do something like:
+
+	Struct<StructLayout> get_someting() const;
+
+	// We know we want to eventually do things like like this exposed to the binder.
+
+	TypedArray<Struct<PropertyInfoLayout>> get_property_list();
+
+	// So what are Struct and StructLayout?
+
+	//We would like to do PropertyInfoLayout like this:
+
+
+	STRUCT_LAYOUT( ProperyInfo, STRUCT_MEMBER("name", Variant::STRING), STRUCT_MEMBER("type", Variant::INT), STRUCT_MEMBER("hint", Variant::INT), STRUCT_MEMBER("hint_string", Variant::STRING), STRUCT_MEMBER("class_name", Variant::STRING) );
+
+	// How does this convert to C?
+
+	// Here is a rough sketch
+	struct StructMember {
+	StringName name;
+	Variant:Type type;
+	StringName class_name;
+	Variant default_value;
+
+	StructMember(const StringName& p_name, const Variant::Type p_type,const Variant& p_default_value = Variant(), const StringName& p_class_name = StringName()) { name = p_name; type=p_type; default_value = p_default_value; class_name = p_class_name; }
+	};
+
+// Important so we force SNAME to it, otherwise this will be leaked memory
+#define STRUCT_MEMBER(m_name,m_type,m_default_value) StructMember(SNAME(m_name),m_type,m_default_value)
+#define STRUCT_CLASS_MEMBER(m_name,m_class) StructMember(SNAME(m_name),Variant::OBJECT,Variant(),m_class)
+
+
+	// StructLayout should ideally be something that we can define like
+
+
+#define STRUCT_LAYOUT(m_class,m_name,...) \
+struct m_name { \
+        _FORCE_INLINE_ static  StringName get_class() { return SNAME(#m_class)); }
+	_FORCE_INLINE_ static  StringName get_name() { return SNAME(#m_name)); }
+	static constexpr uint32_t member_count = GET_ARGUMENT_COUNT;\
+	_FORCE_INLINE_ static const StructMember& get_member(uint32_t p_index) {\
+	CRASH_BAD_INDEX(p_index,member_count)\
+	static StructMember members[member_count]={ __VA_ARGS__ };\
+	return members[p_index];\
+	}\
+};
+
+// Note GET_ARGUMENT_COUNT is a macro that we probably need to add tp typedefs.h, see:
+// https://stackoverflow.com/questions/2124339/c-preprocessor-va-args-number-of-arguments
+
+// Okay, so what is Struct<> ?
+
+// Its a similar class to TypedArray
+
+
+template <class T>
+class Struct : public Array {
```

Comment:
Struct inheriting from Array is kind of awkward. There are many Array methods that need to be manually disabled for Struct (such as `push_back`) and several Array methods that just don't make sense for Struct (like `map`). I understand the desire to not make another `Variant::Type`, as this would have implications for the whole codebase. I wonder whether it would make sense for Struct to not inherit from Array, but still call itself a `Variant::Type::ARRAY` as far as the rest of the codebase is concerned.